---
title: "Multivariate plots for bioinformatics data using ggplot2 and GGally"
author: "Matthew L Bendall"
date: "July 9, 2015"
output: html_document
---

### Load data

```{r setup-workspace}
rm(list=ls())
setwd("~/Projects/UW_institute/SISG8/SISBID.git")


# Load the data
setwd("nprot.data")
samples <- read.table("samples.csv", sep=";", header=TRUE)
dim(samples)
counts = read.csv("counts.csv")
dim(counts)

# source("http://bioconductor.org/biocLite.R")
# biocLite("edgeR")
library(edgeR)
counts <- readDGE(samples$countf)$counts
dim(counts)
setwd("..")
```


### Filter weakly expressed and noninformative (e.g., non-aligned) features

```{r}
noint <- rownames(counts) %in%
  c("no_feature","ambiguous","too_low_aQual",
    "not_aligned","alignment_not_unique")
cpms <- cpm(counts)
counts.all <- counts[,order(samples$condition)]
colnames(counts.all) <- samples$shortname[order(samples$condition)]
head(counts.all)
```

### Keep only genes with at least 1 read per million, for the smallest number of replicates in  treatment

```{r}
keep <- rowSums(cpms > 1) >= 3 & !noint
counts <- counts.all[keep,]
#colnames(counts) = samples$shortname
head(counts, 5)
```

### Create DGEList

```{r}
d <- DGEList(counts = counts, group = samples$condition[order(samples$condition)])
str(d)
```

### Normalize
```{r}
d <- calcNormFactors(d)
str(d)
d$samples
```

### Make MDS plot

Primary summary of an RNA-seq 

```{r}
par(pty="s")
d.mds <- plotMDS(d, labels = samples$shortname[order(samples$condition)],
       col = c("darkgreen","blue")[factor(samples$condition[order(samples$condition)])],
       gene.selection = "common", xlim=c(-1.1, 1.1))
```


### Check the normalization

It is important for the overall distribution from one sample to another to have the same distribution.
Assuming that majority of genes are the same between treatments, with only a few with changes.

#### Make boxplot for each treatment

`geom_boxplot()` could be used, `ggparcood()` puts the data frame into long format.

```{r}
library(ggplot2)
library(GGally)
d$counts <- data.frame(d$counts) # Needs to be a data frame
d$lcounts <- log(d$counts+1)
ggparcoord(d$lcounts, columns=1:7, boxplot=TRUE, scale="globalminmax",
           showPoints=FALSE, alphaLines=0) +
  xlab("") + ylab("log(cpm)") +
  theme_bw()
# ggparcoord(d$lcounts, columns=1:7, alphaLines=0.1)
```

#### Use trimmed mean normalization

```{r}
d <- calcNormFactors(d, method="TMM")
nc <- data.frame(cpm(d, normalized.lib.sizes = TRUE, log=TRUE))
ggparcoord(nc, columns=1:7, boxplot=TRUE, scale="globalminmax", showPoints=FALSE,
           alphaLines=0) +
  xlab("") + ylab("log(cpm)") +
  theme_bw()
```

#### Check that the MDS was conducted on the normalized data

The plot here is the same as the above, so plotMDS is performed on the normalized data.

```{r}
d.mds2 <- plotMDS(nc, labels = samples$shortname[order(samples$condition)],
       col = c("darkgreen","blue")[factor(samples$condition[order(samples$condition)])],
       gene.selection = "common", xlim=c(-1.1, 1.1))
```

#### Scatterplot matrix of normalized data
```{r}
ggscatmat(nc)
```

#### Full parallel coordinate plot

Samples are parallel (vertical).
Values for each gene are connected across samples.

```{r}
ggparcoord(nc, columns=1:7, scale="globalminmax", alphaLines=0.5) +
  xlab("") + ylab("log(cpm)") +
  theme_bw()
ggparcoord(nc, columns=1:7, scale="globalminmax", alphaLines=0.01) +
  xlab("") + ylab("log(cpm)") +
  theme_bw()
```

#### Porcupine plot

Compare two treatments showing both replicates. Line is drawn between replicates.

```{r}
ggplot(nc) + geom_segment(aes(x=CT.PA.1, xend=CT.PA.2, y=KD.PA.3, yend=KD.PA.4)) +
  theme_bw() + theme(aspect.ratio=1)
```

### Check the mean and variance relationship
For the Poisson model, the mean = variance.
RNA-seq tends to be more overdispersed, variance is larger than expected.
Over-dispersed leads to fitting a negative binomial model.

##### Common dispersion
Common dispersion would be used if all genes assumed to have same variance.

```{r}
d <- estimateCommonDisp(d)
d$common.dispersion
```

#### Tagwise dispersion is a weighted average of individual gene dispersion with common dispersion

Tagwise dispersion borrows strength from all genes in dataset.

```{r}
d <- estimateTagwiseDisp(d)
summary(d$tagwise.dispersion)
d$prior.n
mv <- binMeanVar(d, group = d$samples$group)
qplot(mv$means, mv$vars, alpha=I(0.5)) +
  scale_y_log10() + scale_x_log10() +
  geom_smooth(method='lm') + theme_bw() +
  theme(aspect.ratio=1)
```

Low count data

```{r}
qplot(mv$vars, d$tagwise.dispersion) +
  scale_y_log10() + scale_x_log10() +
  geom_smooth() + theme_bw() +
  theme(aspect.ratio=1)
```

```{r}
plotMeanVar(d, show.tagwise.vars = TRUE, NBline = TRUE)
plotBCV(d)
```

```{r}
# Examine the dispersions, in relation to the means
mv <- binMeanVar(d, group = d$samples$group)
# $means are the means for each gene
# $vars are the pooled variances for each gene
qplot(mv$means, mv$vars, alpha=I(0.5)) + scale_x_log10() + scale_y_log10() +
  geom_smooth(method="lm") + theme_bw() + theme(aspect.ratio=1)
qplot(mv$means, d$tagwise.dispersion) + scale_x_log10() + scale_y_log10() +
  geom_smooth() + theme_bw() + theme(aspect.ratio=1)
qplot(mv$vars, d$tagwise.dispersion) + scale_x_log10() + scale_y_log10() +
  geom_smooth() + theme_bw() + theme(aspect.ratio=1)
```

### Significance testing

```{r}

# Test for differential expression (`classic` edgeR)
de <- exactTest(d, pair = c("CTL","KD"))
tt <- topTags(de, n = nrow(d), sort.by="none")
nc.sig <- data.frame(gene=rownames(nc), nc, tt)
nc.sig$sig05 <- ifelse(nc.sig$FDR < 0.05, "S", "NS")
nc.sig$sig01 <- ifelse(nc.sig$FDR < 0.01, "S", "NS")
```

#### Porcupine plot with significance

###### FDR < 0.05

```{r}
ggplot(nc.sig) + geom_segment(aes(x=CT.PA.1, xend=CT.PA.2, y=KD.PA.3,
                                  yend=KD.PA.4, color=sig05)) +
  scale_color_manual(values=c("S"="red", "NS"="grey90")) +
  xlab("CT") + ylab("KD") +
  theme_bw() + theme(aspect.ratio=1)
```

###### FDR < 0.01

```{r}
ggplot(nc.sig) + geom_segment(aes(x=CT.PA.1, xend=CT.PA.2, y=KD.PA.3,
                                  yend=KD.PA.4, color=sig01)) +
  scale_color_manual(values=c("S"="red", "NS"="grey90")) +
  xlab("CT") + ylab("KD") +
  theme_bw() + theme(aspect.ratio=1)
```

### MV plots
```{r}
nc.sig$sig01 <- factor(nc.sig$sig01, levels=c("NS","S"))
ggscatmat(nc.sig, columns=2:8, color="sig01") +
  scale_color_manual(values=c("S"="red", "NS"="white"))

ggscatmat(nc.sig[nc.sig$sig01=="S",], columns=2:8)

ggparcoord(nc.sig, columns=2:8, scale="globalminmax", alphaLines=0.5,
           groupColumn="sig01") +
  scale_color_manual(values=c("S"="red", "NS"="grey90")) +
  xlab("") + ylab("log(cpm)") +
  theme_bw()

ggparcoord(nc.sig[nc.sig$sig01=="S",], columns=2:8, scale="globalminmax",
           alphaLines=0.5) +
  xlab("") + ylab("log(cpm)") +
  theme_bw()
```


### Interaction plots of top genes

Treatments horizontally, response vertically

```{r}
library(tidyr)
library(dplyr)
nc.sig <- arrange(nc.sig, PValue)
```

#### Most significant gene

```{r}
g1 <- gather(nc.sig[1,2:8])
g1$trt <- substr(g1$key, 1, 2)
g1$read <- substr(g1$key, 4, 5)
g1.mean <- summarise(group_by(g1, trt), m=mean(value))
qplot(trt, value, data=g1, xlab="Treatment", ylab="logCPM", colour=read,
      size=I(5), alpha=I(0.5)) +
  annotate("segment", x=1, xend=2, y=g1.mean$m[1],
           yend=g1.mean$m[2], colour="grey80") +
  ggtitle(nc.sig$gene[1]) +
  theme_bw() + theme(aspect.ratio=1)
```

```{r}
table(nc.sig$sig01)
```

#### The largest p-value that meets threshold
```{r}
g314 <- gather(nc.sig[314,2:8])
g314$trt <- substr(g314$key, 1, 2)
g314$read <- substr(g314$key, 4, 5)
g314.mean <- summarise(group_by(g314, trt), m=mean(value))
qplot(trt, value, data=g314, xlab="Treatment", ylab="logCPM", colour=read,
      size=I(5), alpha=I(0.5)) +
  annotate("segment", x=1, xend=2, y=g314.mean$m[1],
           yend=g314.mean$m[2], colour="grey80") +
  ggtitle(nc.sig$gene[314]) +
  theme_bw() + theme(aspect.ratio=1)
```

### Check effect size

To check the effect size, visually, we are going to scramble the labels, and re-run the significance testing

```{r}
dp <- d
ncp <- nc
dp$samples$group <- c("CTL","KD","CTL","KD","CTL","KD","CTL")
dp$samples$group
dep = exactTest(dp, pair = c("CTL","KD"))
ttp = topTags(dep, n = nrow(dp), sort.by="none")
ncp.sig <- data.frame(gene=rownames(ncp), ncp, ttp)
ncp.sig$sig05 <- ifelse(ncp.sig$FDR < 0.05, "S", "NS")
ncp.sig$sig01 <- ifelse(ncp.sig$FDR < 0.01, "S", "NS")
ncp.sig <- arrange(ncp.sig, PValue)
```

Table for significant/non-significant
```{r}
table(ncp.sig$sig01)
```

#### Top gene after scrambling

```{r}
g1 <- gather(ncp.sig[1,2:8])
g1$trt <- dp$samples$group
g1.mean <- summarise(group_by(g1, trt), m=mean(value))
qplot(trt, value, data=g1, xlab="Treatment", ylab="logCPM",
      size=I(5), alpha=I(0.5)) +
  annotate("segment", x=1, xend=2, y=g1.mean$m[1],
           yend=g1.mean$m[2], colour="grey80") +
  ggtitle(ncp.sig$gene[314]) +
  theme_bw() + theme(aspect.ratio=1)
```

```{r}
g314 <- gather(ncp.sig[314,2:8])
g314$trt <- dp$samples$group
g314.mean <- summarise(group_by(g314, trt), m=mean(value))
qplot(trt, value, data=g314, xlab="Treatment", ylab="logCPM",
      size=I(5), alpha=I(0.5)) +
  annotate("segment", x=1, xend=2, y=g314.mean$m[1],
           yend=g314.mean$m[2], colour="grey80") +
  ggtitle(ncp.sig$gene[314]) +
  theme_bw() + theme(aspect.ratio=1)
```
